File: frontend/src/types/reservation.ts
import type { ProductId } from "@/types/product";

export type ReservationStep = "cart" | "details" | "confirm";

export type ReservationStatus = "draft" | "awaiting_whatsapp";

export type CartItem = {
  productId: ProductId;
  quantity: number;
};

export type CustomerDetails = {
  name: string;
  phone: string;
  pickupDate: string;
  pickupTime: string;
  notes: string;
};

export type ReservationDraft = {
  items: CartItem[];
  customer: CustomerDetails;
  status: ReservationStatus;
  step: ReservationStep;
  updatedAt: string;
};

File: backend/app/schemas/reservation.py
from datetime import date
from typing import List, Literal, Optional

from pydantic import BaseModel, Field


ReservationStatus = Literal[
    "Pending",
    "Confirmed",
    "Preparing",
    "Ready",
    "Collected",
    "Cancelled",
    # Legacy alias still accepted on write (normalized to Ready)
    "Ready for Pickup",
]


class CustomerInfo(BaseModel):
    name: str = Field(..., min_length=2)
    phone: str = Field(..., min_length=10)


class ReservationItemCreate(BaseModel):
    product_id: str
    quantity: int = Field(..., ge=1)


class ReservationItem(BaseModel):
    product_id: str
    name: str
    price: int
    quantity: int


class ReservationCreate(BaseModel):
    customer: CustomerInfo
    items: List[ReservationItemCreate] = Field(..., min_length=1)
    pickup_date: date
    pickup_time: str = Field(..., min_length=1)
    notes: Optional[str] = None


class ReservationUpdate(BaseModel):
    """Partial update for customer details, pickup, notes, or items."""

    customer: Optional[CustomerInfo] = None
    items: Optional[List[ReservationItemCreate]] = None
    pickup_date: Optional[date] = None
    pickup_time: Optional[str] = None
    notes: Optional[str] = None
    whatsapp_sent: Optional[bool] = None


class ReservationStatusUpdate(BaseModel):
    status: ReservationStatus
